### PR TITLE
Refactor and cleanup bean context

### PR DESCRIPTION
**Pull request by dstepanov**
This PR removes/refactores some old stuff probably introduced with Micronaut 1 but not used anymore

- Refactored `AbstractInitializableBeanDefinition` to have precalculated fields excluded  for better maintainability
- Removed old bean definition abstract classes
- Introduced `ObjectUtils.hash` as an alternative to `Objects.hash`, which doesn't allocate arrays (WDYT @melix ?).
- Removed unused info about reflection being required for the inject points
- Removed a way to reflectively invoke injection points, that wasn't used anymore
- Removed a fallback way to create a bean definition by getting the constructor and using reflection, that scenario is not used anywhere anyway
- Introduced replacement for `BeanFactory` and `ParametrizedBeanFactory`: `InitializingBeanDefinition`, `ParametrizedInstantiatableBeanDefinition` API now makes more sense and doesn't require to pass the same instance of `BeanDefinition`
- Extracted inject method from `BeanDefinition` to `InjectableBeanDefinition`
- Removed deprecated stuff

### CODE REVIEW

**Comment by graemerocher** [comment]
tests broken

**Comment by graemerocher** [comment]
still failing

**Comment by dstepanov** [comment]
Passing now

File: inject/src/main/java/io/micronaut/inject/InjectableBeanDefinition.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import io.micronaut.context.DefaultBeanResolutionContext;
+
+/**
+ * <p>An type of {@link BeanDefinition} that supports post initialization bean dependencies injection.</p>
+ *
+ * @param <T> The bean type
+ * @author Denis Stepanov
+ * @since 4.0
+ */
+public interface InjectableBeanDefinition<T> extends BeanDefinition<T> {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/InjectableBeanDefinition.java:32]: annotate internal IMO

File: inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import io.micronaut.context.exceptions.BeanInstantiationException;
+
+/**
+ * <p>An type of {@link BeanDefinition} that can build a new instance.</p>
+ *
+ * @param <T> The bean type
+ * @author Denis Stepanov
+ * @since 4.0
+ */
+public interface InstantiatableBeanDefinition<T> extends BeanDefinition<T> {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java:33]: annotate internal IMO

File: inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Builds a bean instance.
+     *
+     * @param context    The context
+     * @return The instance
+     * @throws BeanInstantiationException if the instance could not be instantiated
+     */
+    default T instantiate(BeanContext context) throws BeanInstantiationException {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java]: nullability annotations

File: inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Builds a bean instance.
+     *
+     * @param resolutionContext The bean resolution context
+     * @param context           The context
+     * @return The instance
+     * @throws BeanInstantiationException if the instance could not be instantiated
+     */
+    T instantiate(BeanResolutionContext resolutionContext, BeanContext context) throws BeanInstantiationException;
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java]: nullability annotations

File: inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Builds a bean instance.
+     *
+     * @param context    The context
+     * @return The instance
+     * @throws BeanInstantiationException if the instance could not be instantiated
+     */
+    default T instantiate(BeanContext context) throws BeanInstantiationException {
+        return instantiate(new DefaultBeanResolutionContext(context, this), context);
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/InstantiatableBeanDefinition.java]: should be wrapped in try with resources to that inject scoped beans are destroyed

File: inject/src/main/java/io/micronaut/inject/ParametrizedInstantiatableBeanDefinition.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Map;
+
+/**
+ * <p>An type of {@link BeanDefinition} that can build a new instance, construction requires additional (possibly user supplied) parameters in order construct a bean</p>
+ *
+ * @param <T> The bean type
+ * @author Denis Stepanov
+ * @since 4.0
+ */
+public interface ParametrizedInstantiatableBeanDefinition<T> extends InstantiatableBeanDefinition<T> {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/ParametrizedInstantiatableBeanDefinition.java:35]: nullability annotations

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Review by graemerocher — APPROVED** [review]

**Pull request merged**